### PR TITLE
Remove dependency on BAC0 and bacpypes

### PR DESCRIPTION
**Pull request by piotrbulinski**

### CODE REVIEW

**Comment by piotrbulinski** [comment]
@lellky please take a look at this one. I still need to do the complete testing of the new implementation and add async support, but it's getting where I want it to be (i.e. without any dependencies).

File: flexit_bacnet/nordic.py [hunk 1]
```diff
@@ -2,21 +2,21 @@
...
 """
-from .device_property import DeviceProperty
+from .bacnet import DeviceProperty, ObjectType
 
 # Comfort button [RW]
 # 0 = Ventilation mode Away after Away delay timer duration [Pintval,318].
 #     Also overrides Room operating mode PRESENT_VENTILATION_MODE.
 # 1 = Ventilation mode according to Room operating mode PRESENT_VENTILATION_MODE.
-COMFORT_BUTTON = DeviceProperty('binaryValue', 50, priority=13)
+COMFORT_BUTTON = DeviceProperty(ObjectType.BINARY_VALUE, 50, priority=13)
```
Comment (lellky) [inline flexit_bacnet/nordic.py:11]: This is really good.

File: README.md [hunk 2]
```diff
@@ -45,29 +51,49 @@ For list of available states and interactions, please study [device.py](./flexit
...
\ No newline at end of file
+```
+
+
+## Examples
+
+To execute examples without installing the package, set PYTHONPATH to local directory, e.g.:
+
+```bash
+PYTHONPATH=. python3 examples/current_mode.py 192.168.0.100
```
Comment (lellky) [inline README.md:96]: Setting PYTHONPATH seems to work. Did not not this was possible :)

File: examples/current_mode.py [hunk 3]
```diff
@@ -1,15 +1,29 @@
...
 # import FlexitBACnet
-from flexit_bacnet import (
-    FlexitBACnet,
-)
+from flexit_bacnet import FlexitBACnet, VENTILATION_MODE
+
+
+async def main():
+    if len(sys.argv) < 2:
+        print(f"usage ./{sys.argv[0]} <flexit-unit-ip-address>")
```
Comment (lellky) [inline examples/current_mode.py:10]: Good to have the IP as a parameter!

File: examples/device_info.py [hunk 4]
```diff
@@ -1,14 +1,77 @@
...
     if not device.is_valid():
-        raise Exception('not a valid device')
+        raise Exception("not a valid device")
 
     # check device name and s/n
-    print('Device Name:', device.device_name)
-    print('Serial Number:', device.serial_number)
+    print(f"Device Name: {device.device_name}")
+    print(f"Serial Number: {device.serial_number}")
+    print(f"Outside air temp.: {device.outside_air_temperature} °C")
```
Comment (lellky) [inline examples/device_info.py:27]: This is really nice!

File: flexit_bacnet/bacnet.py [hunk 5]
```diff
@@ -1,73 +1,468 @@
...
+
+        object_type = ObjectType(data >> TAG_OBJECT_TYPE_SHIFT)
+        instance_number = data & 0x3FFFFF
+
+        return object_type, instance_number
+
+    def parse_list_of_results(self) -> ObjectProperties:
+        tag_number, tag_type = self.read_context_tag()
+        if tag_number != 1 or tag_type != 6:
+            raise Exception("expected openning tag")
```
Comment (lellky) [inline flexit_bacnet/bacnet.py:293]: Typo, should be opening

File: flexit_bacnet/bacnet.py [hunk 6]
```diff
@@ -1,73 +1,468 @@
...
+
+            results.append((read_value, value))
+
+        return results
+
+    def read_value(self) -> Any:
+        # check the openning tag
+        tag_number, tag_type = self.read_context_tag()
+        if tag_number != 4 or tag_type != 6:
+            raise Exception("expected openning tag")
```
Comment (lellky) [inline flexit_bacnet/bacnet.py:319]: Same here

**Review by lellky — COMMENTED** [review]
I have looked at your code and it looks OK.

Trying to run the example files (on MacOS with Python 3.10.9 in a venv) gives me
```
(venv) lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/device_info.py 192.168.2.14
Traceback (most recent call last):
  File "/Users/lellky/Projects/flexit_bacnet/examples/device_info.py", line 77, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/lellky/Projects/flexit_bacnet/examples/device_info.py", line 18, in main
    await device.update()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/device.py", line 37, in update
    self._state = await self.bacnet.read_multiple(device_properties)
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 462, in read_multiple
    return _parse_read_property_multiple_response(response)
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 213, in _parse_read_property_multiple_response
    device_state[object_id] = decoder.parse_list_of_results()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 309, in parse_list_of_results
    value = self.read_value()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 319, in read_value
    raise Exception("expected openning tag")
Exception: expected openning tag
```

**Comment by lellky** [comment]
I will try to run it on my Ubuntu machine as soon as I can. Probably tomorrow.

**Comment by ainarspastars** [comment]
Tried running and got the same result as @lellky 
```
Exception: expected openning tag
```
This is with Flexit Nordic S3 unit.

**Comment by lellky** [comment]
I tried debugging the code and commenting out
```
 # @property
    # def room_1_humidity(self) -> float:
   ...

 # @property
    # def room_2_humidity(self) -> float:
    ...

@property
    def room_3_humidity(self) -> float:
    ...
```

and

```
# print(f"Room 1 humidity: {device.room_1_humidity}%")
# print(f"Room 2 humidity: {device.room_2_humidity}%")
# print(f"Room 3 humidity: {device.room_3_humidity}%")
```

and

```
# Humidity sensorts
# ROOM_1_HUMIDITY = DeviceProperty(ObjectType.ANALOG_VALUE, 2093)
# ROOM_2_HUMIDITY = DeviceProperty(ObjectType.ANALOG_VALUE, 2094)
# ROOM_3_HUMIDITY = DeviceProperty(ObjectType.ANALOG_VALUE, 2095)
```

makes it work:

```
(venv) lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/device_info.py "192.168.2.14"
Device Name: HvacFnct21y_A
Serial Number: 800130-002967
Outside air temp.: 4.3 °C
Supply air temp.: 15.5 °C
Extract air temp.: 18.9 °C
Exhaust air temp.: 7.8 °C
Room air temp.: 16.0 °C
Comfort button state: 0
Operation mode: Away
Ventilation mode: Away
Air temp. setpoint Away: 16.0 °C
Air temp. setpoint Home: 21.0 °C
Fireplace duration remaining: 10.0 minutes
Rapid ventilation duration remaining: 10.0 minutes
Supply air fan control signal: 42%
Supply air fan RPM: 1570
Exhaust air fan control signal: 40%
Exhaust air fan RPM: 1480
Electric heater enabled: False
Electric heater nominal power: 0.800000011920929 kW
Electric heater power consumption: 0.0 kW
Fan setpoint - supply air Home: 74%
Fan setpoint - supply air Away: 42%
Fan setpoint - supply air High: 100%
Fan setpoint - supply air Cooker: 50%
Fan setpoint - supply air Fireplace: 90%
Fan setpoint - extract air Home: 70%
Fan setpoint - extract air Away: 40%
Fan setpoint - extract air High: 100%
Fan setpoint - extract air Cooker: 50%
Fan setpoint - extract air Fireplace: 50%
Air filter operating time: 396.0 hours
Air filter exchange interval: 8784.0 days
Air filter polluted: False
Heat-exchanger efficiency: 76%
Heat-exchanger speed: 53%
```

File: flexit_bacnet/device.py [hunk 7]
```diff
@@ -1,16 +1,14 @@
...
 
 
 class FlexitBACnet:
-    def __init__(self, device_address: str, device_id: int):
-        self.device_address = device_address
+    def __init__(self, device_address: str, device_id: int, port: int = bacnet.DEFAULT_BACNET_PORT):
+        self.bacnet = bacnet.BACnetClient(device_address, port)
         self.device_id = device_id
-        self._state: DeviceState | None = None
+        self._state: Optional[bacnet.DeviceState] = None
```
[the remark this answers is not in the document]
Comment (piotrbulinski) [reply flexit_bacnet/device.py:16]: @lellky `Optional` is just a type hint, `= None` is the value assignment, so we need both.

File: flexit_bacnet/bacnet.py [hunk 8]
```diff
@@ -1,73 +1,468 @@
...
+    return bvlc + NPDU + apdu
+
+
+# _parse_read_property_multiple_response and return DeviceState
+def _parse_read_property_multiple_response(response: bytes) -> DeviceState:
+    bvlc_type, bvlc_function, bvlc_length = unpack("!BBH", response[0:4])
+    if bvlc_type != BVLC_TYPE or bvlc_function != BVLC_FUNCTION:
+        raise ConnectionError("unexpected response")
+
+    apdu = response[BVLC_LENGTH + len(NPDU) :]
```
[the remark this answers is not in the document]
Comment (piotrbulinski) [reply flexit_bacnet/bacnet.py]: this should be fixed now

File: flexit_bacnet/bacnet.py (hunk 6 above)
Comment (piotrbulinski) [reply flexit_bacnet/bacnet.py]: fixed

File: flexit_bacnet/bacnet.py (hunk 5 above)
Comment (piotrbulinski) [reply flexit_bacnet/bacnet.py]: fixed

**Comment by piotrbulinski** [comment]
@lellky @ainarspastars can you please try again? I've found an issue with parsing of tags inside BACnet response which should be now resolved. I also added logging of the response on exceptions, so when it fails, I can look at the response and figure out exactly why it failed.

**Comment by lellky** [comment]
Great work. It still does not work with the room humidity sensors. Maybe try with an if statement somewhere (like if model xyz then read humidity sensors).

```
(venv) lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/device_info.py "192.168.2.14"
Traceback (most recent call last):
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 471, in read_multiple
    return _parse_read_property_multiple_response(response)
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 221, in _parse_read_property_multiple_response
    device_state[object_id] = decoder.parse_list_of_results()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 317, in parse_list_of_results
    value = self.read_value()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 339, in read_value
    tag_number, tag_type = self.read_context_tag()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 273, in read_context_tag
    raise DecodingError(f"expected context specific tag, got {tag_class}")
flexit_bacnet.bacnet.DecodingError: expected context specific tag, got 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/lellky/Projects/flexit_bacnet/examples/device_info.py", line 77, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/lellky/Projects/flexit_bacnet/examples/device_info.py", line 18, in main
    await device.update()
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/device.py", line 37, in update
    self._state = await self.bacnet.read_multiple(device_properties)
  File "/Users/lellky/Projects/flexit_bacnet/flexit_bacnet/bacnet.py", line 473, in read_multiple
    raise Exception(f"response decoding failed: {exc}\n{response.hex()}") from exc
Exception: response decoding failed: expected context specific tag, got 0
810a02d1010030010e0c014000321e29554e91014f1f0c0c00013e1e29554e21004f1f0c04c001691e29554e21034f1f0c04c0002a1e29554e21034f1f0c008007c11e29554e44418000004f1f0c008007ca1e29554e4441a800004f1f0c04c001681e29554e21014f1f0c0c00010e1e29554e210a4f1f0c008007f61e29554e44412000004f1f0c04c001651e29554e21014f1f0c0c0001251e29554e211e4f1f0c008007ef1e29554e4441f000004f1f0c000000011e29554e444080f5c34f1f0c000000041e29554e444186cccd4f1f0c000000051e29554e44453360004f1f0c0000000b1e29554e4440f851ec4f1f0c0000000c1e29554e44452800004f1f0c0000003b1e29554e44419aa3d74f1f0c0000004b1e29554e44417fae144f1f0c004000031e29554e44429400004f1f0c004000041e29554e44428be6674f1f0c004000001e29554e4442c800004f1f0c008007e71e29554e4442973de94f1f0c014001bd1e29554e91004f1f0c008000be1e29554e443f4ccccd4f1f0c008000c21e29554e44000000004f1f0c014001921e29554e91004f1f0c0080072b1e29554e4442c800004f1f0c0080072c1e29554e44429400004f1f0c0080072d1e29554e44422800004f1f0c0080072e1e29554e4442b400004f1f0c0080072f1e29554e44424800004f1f0c008007301e29554e4442c800004f1f0c008007311e29554e44428c00004f1f0c008007321e29554e44422000004f1f0c008007331e29554e44424800004f1f0c008007341e29554e44424800004f1f0c0080011d1e29554e4443ca00004f1f0c0080011e1e29554e44460940004f1f0c0140020a1e29554e91004f1f0c04c002651e29554e21014f1f0c0080082d1e29555e9101911f5f1f0c0080082e1e29555e9101911f5f1f0c0080082f1e29555e9101911f5f1f0c020000021e294d4e750e0048766163466e63743231795f414f291c4e750e003830303133302d3030323936374f1f
```

**Comment by piotrbulinski** [comment]
@lellky I added support for property access errors, so this should work for you now.

**Comment by lellky** [comment]
It works! 🎉 

```
(venv) lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/current_mode.py "192.168.2.14"
ventilation mode: Home
(venv) lellky@Jonass-MacBook-Air flexit_bacnet % PYTHONPATH=. python3 examples/device_info.py "192.168.2.14"
Device Name: HvacFnct21y_A
Serial Number: 800130-002967
Outside air temp.: 3.3 °C
Supply air temp.: 16.6 °C
Extract air temp.: 19.3 °C
Exhaust air temp.: 7.0 °C
Room air temp.: 16.1 °C
Room 1 humidity: 0%
Room 2 humidity: 0%
Room 3 humidity: 0%
Comfort button state: 1
Operation mode: Home
Ventilation mode: Home
Air temp. setpoint Away: 16.0 °C
Air temp. setpoint Home: 21.0 °C
Fireplace duration remaining: 10.0 minutes
Rapid ventilation duration remaining: 10.0 minutes
Supply air fan control signal: 74%
Supply air fan RPM: 2816
Exhaust air fan control signal: 69%
Exhaust air fan RPM: 2632
Electric heater enabled: False
Electric heater nominal power: 0.800000011920929 kW
Electric heater power consumption: 0.0 kW
Fan setpoint - supply air Home: 74%
Fan setpoint - supply air Away: 42%
Fan setpoint - supply air High: 100%
Fan setpoint - supply air Cooker: 50%
Fan setpoint - supply air Fireplace: 90%
Fan setpoint - extract air Home: 70%
Fan setpoint - extract air Away: 40%
Fan setpoint - extract air High: 100%
Fan setpoint - extract air Cooker: 50%
Fan setpoint - extract air Fireplace: 50%
Air filter operating time: 408.0 hours
Air filter exchange interval: 8784.0 days
Air filter polluted: False
Heat-exchanger efficiency: 77%
Heat-exchanger speed: 100%
```

**Review by lellky — APPROVED** [review]

**Pull request merged**